src/Hub.cpp

```diff
@@ -260,6 +267,38 @@ void Hub::AssignVespeneHarvester(const sc2::Unit& refinery_) {
     worker->GatherVespene(refinery_);
 }
 
+void Hub::DistrubuteMineralWorker(const sc2::Unit* unit_) {
+    if (!unit_)
```

Comment:
The actual distribution should be part of the Miner subsystem. Hub shouldn't be too smart as it is barely an interface over cache. And actually it'll be reworked in a way to provide cached data.

Reply:
Currently, the Miner subsystem cannot be talked to directly by Strategy. It has to go through either the Builder (passed in OnStep, etc) or Hub (global gHub). Either Builder or Hub will have to have lots of pass-thru functions to save and get information, or Hub/Builder will need a pointer to the Miner object.

Reply:
It doesn't need to talk to strategy as strategy can pass some settings/hints into the miner during plugins creation.
SO the question is: what data miner needs to do workers distribution according to the strategy?

Reply:
Doing it during plugins creation would be messy, as Dispatcher won't know which info needs to be passed, so if we need to pass more than one variable, it will clog up plugins creation. Currently, Miner is before Strategy in m_plugins, and all m_plugins call OnGameStart in a for loop. Calling Strategy explicitly before the For loop, or Miner after the For loop, and then passing the info manually seems to create more complication then just using get/set in Hub, which is more flexible, as after plugins creation, then you cannot change the max_workers in reaction to how the game is progressing.

Reply:
The other option may be to use a PubSub method that plugins can check so they don't need explicit pointers to each other. I think there will be more need for that later on as Strategies may want to change many default behaviors in other plugins.

Reply:
Well, it is not so scary as it seems.

a. We definitely need to know which strategy to use in advance. As a result, the strategy should always be created before any other plugins.
b. The strategy could provide a kind of requirements (another struct) with some settings. Those settings could include different hints for different subsystems (in or case for the miner). The things like worker type (i.e. race specific) could be included into the settings, too.
c. Dispatcher takes a const reference to the settings from the strategy.
d. Dispatcher passes the settings by reference into each plugin. A plugin in its turn can put the hints it needs into member variables (also constants).

Such approach allows to pass hints not only to the miner, but to any other plugin and keeps cache role on the hub, not game brain :)

Reply:
I agree, eventually, we need a kind of wire (channel) for the subsystems to communicate using some kind of events, but this is relatively complex task, definitely not for this PR.

Reply:
Also, speaking about more general solutions, it seems feasible to move turns skipping into Dispatcher, i.e.

a. The base Plugin class has a new field turns_to_skip (0 by default) and allows users to get it's value.
b. Subsystems can set turns_to_skip to any value if needed.
d. On each onStep event Dispatcher checks turns_to_skip of each plugin and don't triggers onStep method, if requested.

This approach also allows to give more advanced features for other subsystems.

Reply:
Cool, I like the Settings struct idea. I've made a new Issue for that (#54 "Strategy Settings for plugins"), and removed the get/set in Pub. I'll leave the 80 as-is for now, until #54 is done, if that's alright.

I think everything else in this PR should be resolved now.

Reply:
DistributeMineralWorkers was moved into Miner, so this Comment is resolved.